// ton/dist/index.js
"use strict";
/**
 * Copyright (c) Whales Corp.
 * All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.computeStorageFees = exports.computeMessageForwardFees = exports.computeGasPrices = exports.computeFwdFees = exports.computeExternalMessageFees = exports.loadConfigParamsAsSlice = exports.loadConfigParamById = exports.parseFullConfig = exports.parseVotingSetup = exports.parseValidatorSet = exports.parseProposalSetup = exports.parseBridge = exports.configParseWorkchainDescriptor = exports.configParseValidatorSet = exports.configParseMsgPrices = exports.configParseMasterAddressRequired = exports.configParseMasterAddress = exports.configParseGasLimitsPrices = exports.configParseBridge = exports.configParse40 = exports.configParse29 = exports.configParse28 = exports.configParse18 = exports.configParse17 = exports.configParse16 = exports.configParse15 = exports.configParse13 = exports.configParse12 = exports.configParse8 = exports.configParse5 = exports.ElectorContract = exports.MultisigWallet = exports.MultisigOrderBuilder = exports.MultisigOrder = exports.JettonWallet = exports.JettonMaster = exports.WalletContractV4 = exports.WalletContractV3R2 = exports.WalletContractV3R1 = exports.WalletContractV2R2 = exports.WalletContractV2R1 = exports.WalletContractV1R3 = exports.WalletContractV1R2 = exports.WalletContractV1R1 = exports.TonClient4 = exports.TonClient = exports.HttpApi = void 0;
__exportStar(require("ton-core"), exports);
//
// toncenter Client
//
var HttpApi_1 = require("./client/api/HttpApi");
Object.defineProperty(exports, "HttpApi", { enumerable: true, get: function () { return HttpApi_1.HttpApi; } });
var TonClient_1 = require("./client/TonClient");
Object.defineProperty(exports, "TonClient", { enumerable: true, get: function () { return TonClient_1.TonClient; } });
//
// API V4 Client
//
var TonClient4_1 = require("./client/TonClient4");
Object.defineProperty(exports, "TonClient4", { enumerable: true, get: function () { return TonClient4_1.TonClient4; } });
//
// Wallets
//
var WalletContractV1R1_1 = require("./wallets/WalletContractV1R1");
Object.defineProperty(exports, "WalletContractV1R1", { enumerable: true, get: function () { return WalletContractV1R1_1.WalletContractV1R1; } });
var WalletContractV1R2_1 = require("./wallets/WalletContractV1R2");
Object.defineProperty(exports, "WalletContractV1R2", { enumerable: true, get: function () { return WalletContractV1R2_1.WalletContractV1R2; } });
var WalletContractV1R3_1 = require("./wallets/WalletContractV1R3");
Object.defineProperty(exports, "WalletContractV1R3", { enumerable: true, get: function () { return WalletContractV1R3_1.WalletContractV1R3; } });
var WalletContractV2R1_1 = require("./wallets/WalletContractV2R1");
Object.defineProperty(exports, "WalletContractV2R1", { enumerable: true, get: function () { return WalletContractV2R1_1.WalletContractV2R1; } });
var WalletContractV2R2_1 = require("./wallets/WalletContractV2R2");
Object.defineProperty(exports, "WalletContractV2R2", { enumerable: true, get: function () { return WalletContractV2R2_1.WalletContractV2R2; } });
var WalletContractV3R1_1 = require("./wallets/WalletContractV3R1");
Object.defineProperty(exports, "WalletContractV3R1", { enumerable: true, get: function () { return WalletContractV3R1_1.WalletContractV3R1; } });
var WalletContractV3R2_1 = require("./wallets/WalletContractV3R2");
Object.defineProperty(exports, "WalletContractV3R2", { enumerable: true, get: function () { return WalletContractV3R2_1.WalletContractV3R2; } });
var WalletContractV4_1 = require("./wallets/WalletContractV4");
Object.defineProperty(exports, "WalletContractV4", { enumerable: true, get: function () { return WalletContractV4_1.WalletContractV4; } });
//
// Jettons
//
var JettonMaster_1 = require("./jetton/JettonMaster");
Object.defineProperty(exports, "JettonMaster", { enumerable: true, get: function () { return JettonMaster_1.JettonMaster; } });
var JettonWallet_1 = require("./jetton/JettonWallet");
Object.defineProperty(exports, "JettonWallet", { enumerable: true, get: function () { return JettonWallet_1.JettonWallet; } });
//
// Multisig
//
var MultisigOrder_1 = require("./multisig/MultisigOrder");
Object.defineProperty(exports, "MultisigOrder", { enumerable: true, get: function () { return MultisigOrder_1.MultisigOrder; } });
var MultisigOrderBuilder_1 = require("./multisig/MultisigOrderBuilder");
Object.defineProperty(exports, "MultisigOrderBuilder", { enumerable: true, get: function () { return MultisigOrderBuilder_1.MultisigOrderBuilder; } });
var MultisigWallet_1 = require("./multisig/MultisigWallet");
Object.defineProperty(exports, "MultisigWallet", { enumerable: true, get: function () { return MultisigWallet_1.MultisigWallet; } });
//
// Elector
//
var ElectorContract_1 = require("./elector/ElectorContract");
Object.defineProperty(exports, "ElectorContract", { enumerable: true, get: function () { return ElectorContract_1.ElectorContract; } });
//
// Config
//
var ConfigParser_1 = require("./config/ConfigParser");
Object.defineProperty(exports, "configParse5", { enumerable: true, get: function () { return ConfigParser_1.configParse5; } });
Object.defineProperty(exports, "configParse8", { enumerable: true, get: function () { return ConfigParser_1.configParse8; } });
Object.defineProperty(exports, "configParse12", { enumerable: true, get: function () { return ConfigParser_1.configParse12; } });
Object.defineProperty(exports, "configParse13", { enumerable: true, get: function () { return ConfigParser_1.configParse13; } });
Object.defineProperty(exports, "configParse15", { enumerable: true, get: function () { return ConfigParser_1.configParse15; } });
Object.defineProperty(exports, "configParse16", { enumerable: true, get: function () { return ConfigParser_1.configParse16; } });
Object.defineProperty(exports, "configParse17", { enumerable: true, get: function () { return ConfigParser_1.configParse17; } });
Object.defineProperty(exports, "configParse18", { enumerable: true, get: function () { return ConfigParser_1.configParse18; } });
Object.defineProperty(exports, "configParse28", { enumerable: true, get: function () { return ConfigParser_1.configParse28; } });
Object.defineProperty(exports, "configParse29", { enumerable: true, get: function () { return ConfigParser_1.configParse29; } });
Object.defineProperty(exports, "configParse40", { enumerable: true, get: function () { return ConfigParser_1.configParse40; } });
Object.defineProperty(exports, "configParseBridge", { enumerable: true, get: function () { return ConfigParser_1.configParseBridge; } });
Object.defineProperty(exports, "configParseGasLimitsPrices", { enumerable: true, get: function () { return ConfigParser_1.configParseGasLimitsPrices; } });
Object.defineProperty(exports, "configParseMasterAddress", { enumerable: true, get: function () { return ConfigParser_1.configParseMasterAddress; } });
Object.defineProperty(exports, "configParseMasterAddressRequired", { enumerable: true, get: function () { return ConfigParser_1.configParseMasterAddressRequired; } });
Object.defineProperty(exports, "configParseMsgPrices", { enumerable: true, get: function () { return ConfigParser_1.configParseMsgPrices; } });
Object.defineProperty(exports, "configParseValidatorSet", { enumerable: true, get: function () { return ConfigParser_1.configParseValidatorSet; } });
Object.defineProperty(exports, "configParseWorkchainDescriptor", { enumerable: true, get: function () { return ConfigParser_1.configParseWorkchainDescriptor; } });
Object.defineProperty(exports, "parseBridge", { enumerable: true, get: function () { return ConfigParser_1.parseBridge; } });
Object.defineProperty(exports, "parseProposalSetup", { enumerable: true, get: function () { return ConfigParser_1.parseProposalSetup; } });
Object.defineProperty(exports, "parseValidatorSet", { enumerable: true, get: function () { return ConfigParser_1.parseValidatorSet; } });
Object.defineProperty(exports, "parseVotingSetup", { enumerable: true, get: function () { return ConfigParser_1.parseVotingSetup; } });
Object.defineProperty(exports, "parseFullConfig", { enumerable: true, get: function () { return ConfigParser_1.parseFullConfig; } });
Object.defineProperty(exports, "loadConfigParamById", { enumerable: true, get: function () { return ConfigParser_1.loadConfigParamById; } });
Object.defineProperty(exports, "loadConfigParamsAsSlice", { enumerable: true, get: function () { return ConfigParser_1.loadConfigParamsAsSlice; } });
//
// Fees
//
var fees_1 = require("./utils/fees");
Object.defineProperty(exports, "computeExternalMessageFees", { enumerable: true, get: function () { return fees_1.computeExternalMessageFees; } });
Object.defineProperty(exports, "computeFwdFees", { enumerable: true, get: function () { return fees_1.computeFwdFees; } });
Object.defineProperty(exports, "computeGasPrices", { enumerable: true, get: function () { return fees_1.computeGasPrices; } });
Object.defineProperty(exports, "computeMessageForwardFees", { enumerable: true, get: function () { return fees_1.computeMessageForwardFees; } });
Object.defineProperty(exports, "computeStorageFees", { enumerable: true, get: function () { return fees_1.computeStorageFees; } });
